### PR TITLE
New audit-npm command

### PR DESCRIPTION
**Pull request by talarian1**
- [ ] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: xray/commands/audit/npm.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	jsonOut, err := json.Marshal(res)
+	print(string(jsonOut))
+	return err
+}
+
+func parseNpmDependenciesList(dependencies map[string]*npm.Dependency, packageInfo *commandsutils.PackageInfo) (xrDependencyTree *services.GraphNode) {
+	treeMap := make(map[string][]string)
+	for dependencyId, dependency := range dependencies {
+		dependencyId = NpmPackageTypeIdentifier + dependencyId
+		// Because we are dealing with a
```
Comment (yahavi) [inline xray/commands/audit/npm.go:120]: ?

File: xray/commands/audit/npm.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		father := NpmPackageTypeIdentifier + dependency.GetPathToRoot()[0][0]
+		if sons, ok := treeMap[father]; ok {
```
Comment (yahavi) [inline xray/commands/audit/npm.go:122]: Let's keep the convention of parent-child
```suggestion
		parent := NpmPackageTypeIdentifier + dependency.GetPathToRoot()[0][0]
		if children, ok := treeMap[father]; ok {
```

File: xray/commands/audit/npm.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			treeMap[father] = append(sons, dependencyId)
+		} else {
+			treeMap[father] = []string{dependencyId}
+		}
+	}
+	return buildXrayDependencyTree(treeMap, NpmPackageTypeIdentifier+packageInfo.BuildInfoModuleId())
+}
+
+func buildXrayDependencyTree(treeHelper map[string][]string, node string) *services.GraphNode {
```
Comment (yahavi) [inline xray/commands/audit/npm.go:131]:
```suggestion
func buildXrayDependencyTree(treeHelper map[string][]string, nodeId string) *services.GraphNode {
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Looks good!

I think that since a lot of the npm code is shared between Artifactory and Xray commands, we should probably export the shared code to a common NpmUtils class. The Xray command should not know the Artifactory command.

File: xray/commands/audit/npm.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
+package scan
```
Comment (yahavi) [inline xray/commands/audit/npm.go:1]:
```suggestion
package audit
```

File: xray/commands/audit/npm_test.go [hunk 5]
```diff
@@ -0,0 +1 @@
+package scan
```
Comment (yahavi) [inline xray/commands/audit/npm_test.go:1]:
```suggestion
package audit
```

**Review by yahavi — CHANGES_REQUESTED** [review]

**Comment by yahavi** [comment]
One more thing - please add an option to read the resolver configuration at the `.jfrog/projects/npm.yaml`. This way we will allow users to download dependencies from Artifactory, which is mandatory IMO.

File: artifactory/commands/npm/installorci.go [hunk 6]
```diff
@@ -118,6 +119,15 @@ func (nca *NpmCommandArgs) SetThreads(threads int) *NpmCommandArgs {
...
+func (nca *NpmCommandArgs) SetTypeRestriction(typeRestriction TypeRestriction) *NpmCommandArgs {
+	nca.typeRestriction = typeRestriction
+	return nca
+}
+
+func (nca *NpmCommandArgs) SetPackageInfo(packageInfo *commandUtils.PackageInfo) *NpmCommandArgs {
+	nca.packageInfo = packageInfo
+	return nca
+}
 func NewNpmCommandArgs(npmCommand string) *NpmCommandArgs {
```
Comment (eyalbe4) [inline artifactory/commands/npm/installorci.go:131]: Missing new-line before function.

File: xray/commands/audit/npm.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/jfrog/jfrog-cli-core/xray/commands"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/xray/services"
+)
+
+const (
+	NpmPackageTypeIdentifier = "npm://"
+)
+
+type XrAuditNpmCommand struct {
```
Comment (eyalbe4) [inline xray/commands/audit/npm.go:19]: XrAuditNpmCommand --> AuditNpmCommand

File: xray/commands/audit/npm.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	serverDetails    *config.ServerDetails
+	workingDirectory string
+	arguments        []string
+	typeRestriction  npm.TypeRestriction
```
Comment (eyalbe4) [inline xray/commands/audit/npm.go:23]: ```npm.TypeRestriction``` is under the ```artifactory/commands``` path, but we're under ```xrat/commands```.
We should move shared resources to be under a common package, which can be shared between Artifactory and Xray.
I think there are other components that need to be moved, such as ```npm.NewNpmCommandArgs```.

File: xray/commands/audit/npm.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+func (auditCmd *XrAuditNpmCommand) SetServerDetails(server *config.ServerDetails) *XrAuditNpmCommand {
+	auditCmd.serverDetails = server
+	return auditCmd
+}
+
+func (auditCmd *XrAuditNpmCommand) ServerDetails() (*config.ServerDetails, error) {
+	return auditCmd.serverDetails, nil
+}
+
+func NewXrAuditNpmCommand() *XrAuditNpmCommand {
```
Comment (eyalbe4) [inline xray/commands/audit/npm.go:50]: NewXrAuditNpmCommand --> NewAuditNpmCommand

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: xray/commands/audit/npm.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	if err != nil {
+		return err
+	}
+
+	scanResults, err := xrayManager.GetScanGraphResults(scanId)
+	if err != nil {
+		return err
+	}
+	return printTable(scanResults)
+
```
Comment (eyalbe4) [inline xray/commands/audit/npm.go:107]: Redundant new-line.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Comment by talarian1** [comment]
#143 includes those changes.

**Pull request closed without merging**